File: core/src/main/java/com/pmcl/core/install/VersionStaging.java
package com.pmcl.core.install;

import com.pmcl.core.util.FileUtils;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;

/**
 * 版本目录暂存 / 原子提升工具。
 * <p>
 * 安装过程中写入 {@code versions/{id}.staging/}，成功后再提升为 {@code versions/{id}/}，
 * 避免半成品被 {@link com.pmcl.core.version.VersionManager} 扫描为可启动版本。
 */
public final class VersionStaging {

    public static final String STAGING_SUFFIX = ".staging";
    public static final String BAK_SUFFIX = ".bak";

    private VersionStaging() {}

    /** 是否为暂存或回滚备份目录名。 */
    public static boolean isTransientDirName(String name) {
        return name != null && (name.endsWith(STAGING_SUFFIX) || name.endsWith(BAK_SUFFIX));
    }

    /**
     * 拒绝路径穿越式版本 id（远程 profile.id / 用户输入）。
     */
    public static void assertSafeVersionId(String versionId) throws IOException {
        if (versionId == null || versionId.isBlank()) {
            throw new IOException("版本 id 为空，拒绝写入");
        }
        if (versionId.indexOf('\0') >= 0
                || versionId.contains("/")
                || versionId.contains("\\")
                || versionId.contains("..")
                || versionId.equals(".")
                || versionId.equals("..")) {
            throw new IOException("非法版本 id（含路径穿越字符）: " + versionId);
        }
    }

    public static Path stagingDir(Path versionsDir, String versionId) throws IOException {
        assertSafeVersionId(versionId);
        Path base = versionsDir.toAbsolutePath().normalize();
        Path staging = base.resolve(versionId + STAGING_SUFFIX).normalize();
        if (!staging.startsWith(base)) {
            throw new IOException("版本 staging 路径越界: " + versionId);
        }
        return staging;
    }

    /**
     * 将版本 JSON 写入 staging 目录（文件名仍为 {@code {id}.json}）。
     *
     * @return staging 目录路径
     */
    public static Path writeVersionJson(Path versionsDir, String versionId, String jsonContent)
            throws IOException {
        Path staging = stagingDir(versionsDir, versionId);
        Files.createDirectories(staging);
        Path jsonFile = staging.resolve(versionId + ".json").normalize();
        if (!jsonFile.startsWith(staging.toAbsolutePath().normalize())) {
            throw new IOException("版本 JSON 路径越界: " + versionId);
        }
        Files.writeString(jsonFile, jsonContent, StandardCharsets.UTF_8);
        return staging;
    }

    /**
     * 将 staging 目录提升为正式版本目录。
     * 若正式目录已存在，先移到 {@code .bak}，成功后再删除备份。
     */
    public static void promote(Path versionsDir, String versionId, Path stagingDir)
            throws IOException {
        assertSafeVersionId(versionId);
        Path base = versionsDir.toAbsolutePath().normalize();
        Path finalDir = base.resolve(versionId).normalize();
        Path bakDir = base.resolve(versionId + BAK_SUFFIX).normalize();
        if (!finalDir.startsWith(base) || !bakDir.startsWith(base)) {
            throw new IOException("版本目录路径越界: " + versionId);
        }
        FileUtils.deleteRecursively(bakDir);

        if (Files.exists(finalDir)) {
            try {
                Files.move(finalDir, bakDir,
                        StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING);
            } catch (java.nio.file.AtomicMoveNotSupportedException e) {
                Files.move(finalDir, bakDir, StandardCopyOption.REPLACE_EXISTING);
            }
        }

        try {
            try {
                Files.move(stagingDir, finalDir, StandardCopyOption.ATOMIC_MOVE);
            } catch (java.nio.file.AtomicMoveNotSupportedException e) {
                Files.move(stagingDir, finalDir);
            }
        } catch (IOException e) {
            if (Files.exists(bakDir) && !Files.exists(finalDir)) {
                try {
                    Files.move(bakDir, finalDir);
                } catch (IOException restoreErr) {
                    e.addSuppressed(restoreErr);
                }
            }
            throw new IOException("无法将安装暂存目录提升为正式版本: " + versionId, e);
        }
        FileUtils.deleteRecursively(bakDir);
    }

    /** 删除指定版本的 staging 目录（失败清理用）。 */
    public static void discard(Path versionsDir, String versionId) {
        if (versionId == null || versionId.isBlank()) return;
        try {
            FileUtils.deleteRecursively(stagingDir(versionsDir, versionId));
        } catch (IOException e) {
            // 非法 id：忽略，避免取消路径因坏 id 再抛错
        }
    }
}

File: pcmlAndroid/core/src/main/java/com/lash/pmcl/core/install/VersionStaging.kt
package com.lash.pmcl.core.install

import com.lash.pmcl.core.util.FileUtils
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption

/**
 * 版本目录暂存 / 原子提升工具。
 * <p>
 * 安装过程中写入 `versions/{id}.staging/`，成功后再提升为 `versions/{id}/`，
 * 避免半成品被 VersionManager 扫描为可启动版本。
 */
object VersionStaging {
    const val STAGING_SUFFIX = ".staging"
    const val BAK_SUFFIX = ".bak"

    /** 是否为暂存或回滚备份目录名。 */
    fun isTransientDirName(name: String?): Boolean {
        return name != null && (name.endsWith(STAGING_SUFFIX) || name.endsWith(BAK_SUFFIX))
    }

    /**
     * 拒绝路径穿越式版本 id（远程 profile.id / 用户输入）。
     */
    @Throws(IOException::class)
    fun assertSafeVersionId(versionId: String?) {
        if (versionId.isNullOrBlank()) {
            throw IOException("版本 id 为空，拒绝写入")
        }
        if (versionId.indexOf('\u0000') >= 0
            || versionId.contains("/")
            || versionId.contains("\\")
            || versionId.contains("..")
            || versionId == "."
            || versionId == ".."
        ) {
            throw IOException("非法版本 id（含路径穿越字符）: $versionId")
        }
    }

    @Throws(IOException::class)
    fun stagingDir(versionsDir: Path, versionId: String): Path {
        assertSafeVersionId(versionId)
        val base = versionsDir.toAbsolutePath().normalize()
        val staging = base.resolve(versionId + STAGING_SUFFIX).normalize()
        if (!staging.startsWith(base)) {
            throw IOException("版本 staging 路径越界: $versionId")
        }
        return staging
    }

    /**
     * 将版本 JSON 写入 staging 目录（文件名仍为 `{id}.json`）。
     *
     * @return staging 目录路径
     */
    @Throws(IOException::class)
    fun writeVersionJson(versionsDir: Path, versionId: String, jsonContent: String): Path {
        val staging = stagingDir(versionsDir, versionId)
        Files.createDirectories(staging)
        val jsonFile = staging.resolve("$versionId.json").normalize()
        if (!jsonFile.startsWith(staging.toAbsolutePath().normalize())) {
            throw IOException("版本 JSON 路径越界: $versionId")
        }
        FileUtils.writeString(jsonFile, jsonContent, StandardCharsets.UTF_8)
        return staging
    }

    /**
     * 将 staging 目录提升为正式版本目录。
     * 若正式目录已存在，先移到 `.bak`，成功后再删除备份。
     */
    @Throws(IOException::class)
    fun promote(versionsDir: Path, versionId: String, stagingDir: Path) {
        assertSafeVersionId(versionId)
        val base = versionsDir.toAbsolutePath().normalize()
        val target = base.resolve(versionId).normalize()
        if (!target.startsWith(base)) {
            throw IOException("版本目录路径越界: $versionId")
        }
        // 已有正式目录 → 先备份
        val bak = base.resolve(versionId + BAK_SUFFIX).normalize()
        if (Files.exists(target)) {
            if (Files.exists(bak)) {
                FileUtils.deleteRecursively(bak)
            }
            Files.move(target, bak, StandardCopyOption.REPLACE_EXISTING)
        }
        try {
            Files.move(stagingDir, target, StandardCopyOption.REPLACE_EXISTING)
        } catch (e: Exception) {
            // 提升失败：尝试回滚
            try {
                if (Files.exists(bak)) {
                    Files.move(bak, target, StandardCopyOption.REPLACE_EXISTING)
                }
            } catch (rollbackErr: Exception) {
                e.addSuppressed(rollbackErr)
                System.err.println("[VersionStaging] 回滚也失败: $versionId — ${rollbackErr.message}")
            }
            throw IOException("提升 staging 目录失败: $versionId", e)
        }
        // 成功后删除备份（失败不影响主流程）
        try {
            if (Files.exists(bak)) {
                FileUtils.deleteRecursively(bak)
            }
        } catch (e: IOException) {
            System.err.println("[VersionStaging] 删除备份失败: ${e.message}")
        }
    }

    /** 回滚到备份（若存在） */
    @Throws(IOException::class)
    fun rollback(versionsDir: Path, versionId: String) {
        assertSafeVersionId(versionId)
        val base = versionsDir.toAbsolutePath().normalize()
        val target = base.resolve(versionId).normalize()
        val bak = base.resolve(versionId + BAK_SUFFIX).normalize()
        if (!Files.exists(bak)) return
        if (Files.exists(target)) {
            FileUtils.deleteRecursively(target)
        }
        Files.move(bak, target, StandardCopyOption.REPLACE_EXISTING)
    }

    /** 删除 staging 目录（安装失败时清理） */
    @Throws(IOException::class)
    fun discard(versionsDir: Path, versionId: String) {
        assertSafeVersionId(versionId)
        val base = versionsDir.toAbsolutePath().normalize()
        val staging = base.resolve(versionId + STAGING_SUFFIX).normalize()
        val bak = base.resolve(versionId + BAK_SUFFIX).normalize()
        if (Files.exists(staging)) FileUtils.deleteRecursively(staging)
        if (Files.exists(bak)) FileUtils.deleteRecursively(bak)
    }
}
